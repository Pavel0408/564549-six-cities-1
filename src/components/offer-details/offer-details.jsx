import React, {PureComponent} from "react";
import {Link} from "react-router-dom";
import PropTypes from "prop-types";
import {ReviewsList} from "../reviews-list/reviews-list";
import {WithActiveItem} from "../../hocs/with-active-item";
import {OffersList} from "../offers-list/offers-list";
import {WithLeafletMap} from "../with-leaflet-map/with-leaflet-map";
import {OffersMap} from "../offers-map/offers-map";
import {getDistanceFromCoords} from "../../utils";


export class OfferDetails extends PureComponent {

  getDistanceFromActiveOffer(offer) {
    return getDistanceFromCoords({
      coordinateFirst: this.props.activeOffer.coordinates,
      coordinateSecond: offer.coordinates
    });
  }

  render() {
    const offer = this.props.activeOffer;
    const offers = this.props.offers.filter((offersItem) => offersItem !== offer)
      .sort((a, b) => {
        return this.getDistanceFromActiveOffer(a) - this.getDistanceFromActiveOffer(b);
      }).slice(0, 3);

    const OffersListWithActiveItem = <WithActiveItem
      render={(childProps) => <OffersList {...childProps} offers={offers} changeActiveOffer={this.props.changeActiveOffer}/>}
    />;

    const {isAuthorizationRequired, user} = this.props;
    const userElementSwitch = () => {
      return isAuthorizationRequired && user ?
        <Link
          className="header__nav-link header__nav-link--profile"
          to={`/favorites`}> <div className="header__avatar-wrapper user__avatar-wrapper">
            <img src={`https://es31-server.appspot.com/six-cities${user.avatar}`}/>
            <span className="header__user-name user__name">{user.email}</span>
          </div>
        </Link>
        :
        <Link className="header__nav-link header__nav-link--profile" to={`/login`}>
          <div className="header__avatar-wrapper user__avatar-wrapper">
          </div>
          <span className="header__login">Sign in</span>
        </Link>;
    };

    return <React.Fragment>
      <div>
        <div style={{display: `none`}}>
          <svg xmlns="http://www.w3.org/2000/svg"><symbol id="icon-arrow-select" viewBox="0 0 7 4"><path fillRule="evenodd" clipRule="evenodd" d="M0 0l3.5 2.813L7 0v1.084L3.5 4 0 1.084V0z" /></symbol><symbol id="icon-bookmark" viewBox="0 0 17 18"><path d="M3.993 2.185l.017-.092V2c0-.554.449-1 .99-1h10c.522 0 .957.41.997.923l-2.736 14.59-4.814-2.407-.39-.195-.408.153L1.31 16.44 3.993 2.185z" /></symbol><symbol id="icon-star" viewBox="0 0 13 12"><path fillRule="evenodd" clipRule="evenodd" d="M6.5 9.644L10.517 12 9.451 7.56 13 4.573l-4.674-.386L6.5 0 4.673 4.187 0 4.573 3.549 7.56 2.483 12 6.5 9.644z" /></symbol></svg>
        </div>
        <header className="header">
          <div className="container">
            <div className="header__wrapper">
              <div className="header__left">
                <a className="header__logo-link" href="main.html">
                  <img className="header__logo" src="/img/logo.svg" alt="6 cities logo" width={81} height={41} />
                </a>
              </div>
              <nav className="header__nav">
                <ul className="header__nav-list">
                  <li className="header__nav-item user">
                    {userElementSwitch()}
                  </li>
                </ul>
              </nav>
            </div>
          </div>
        </header>
        <main className="page__main page__main--property">
          <section className="property">
            <div className="property__gallery-container container">

              <div className="property__gallery">
                {offer.allImages.slice(0, 6).map((image, i) =>{
                  return <div key={i.toString() + offer.id} className="property__image-wrapper">
                    <img className="property__image" src={image} alt="Photo studio" />
                  </div>;
                })
                }
              </div>
            </div>
            <div className="property__container container">
              <div className="property__wrapper">
                {offer.isPremium && <div className="property__mark">
                  <span>Premium</span>
                </div>}
                <div className="property__name-wrapper">
                  <h1 className="property__name">
                    {offer.name}
                  </h1>
                  <button className={offer.isFavorite ? `property__bookmark-button  property__bookmark-button--active button` : `property__bookmark-button button`} type="button">
                    <svg className="property__bookmark-icon" width={31} height={33}>
                      <use xlinkHref="#icon-bookmark" />
                    </svg>
                    <span className="visually-hidden">To bookmarks</span>
                  </button>
                </div>
                <div className="property__rating rating">
                  <div className="property__stars rating__stars">
                    <span style={{width: `${offer.rating * 10}%`}} />
                    <span className="visually-hidden">Rating</span>
                  </div>
                  <span className="property__rating-value rating__value">{offer.rating / 2}</span>
                </div>
                <ul className="property__features">
                  <li className="property__feature property__feature--entire">
                    Entire place
                  </li>
                  <li className="property__feature property__feature--bedrooms">
                    {offer.bedrooms} Bedrooms
                  </li>
                  <li className="property__feature property__feature--adults">
                    Max {offer.maxAdults} adults
                  </li>
                </ul>
                <div className="property__price">
                  <b className="property__price-value">â‚¬{this.props.activeOffer.price}</b>
                  <span className="property__price-text">&nbsp;night</span>
                </div>
                <div className="property__inside">
                  <h2 className="property__inside-title">What&apos;s inside</h2>
                  <ul className="property__inside-list">
                    {offer.goods.map((item, i)=>{
                      return <li key={i.toString() + offer.id} className="property__inside-item">
                        {item}
                      </li>;
                    })}
                  </ul>
                </div>
                <div className="property__host">
                  <h2 className="property__host-title">Meet the host</h2>
                  <div className="property__host-user user">
                    <div className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper">
                      <img className="property__avatar user__avatar" src={`/${offer.host.avatar}`} width={74} height={74} alt="Host avatar" />
                    </div>
                    <span className="property__user-name">
                      {offer.host.name}
                    </span>
                    <span className="property__user-status">
                      {offer.host.isPro && `Pro`}
                    </span>
                  </div>
                  <div className="property__description">
                    <p className="property__text">
                      {offer.description}
                    </p>
                  </div>
                </div>
                <ReviewsList {...this.props}/>
              </div>
            </div>
            <section className="property__map map" style={{backgroundImage: `none`}}>
              <WithLeafletMap
                render={(data) => <OffersMap mapMethods={data} offers={offers} cityName={this.props.cityName}/>}/>
            </section>
          </section>
          <div className="container">
            <section className="near-places places">
              <h2 className="near-places__title">Other places in the neighbourhood</h2>
              <div className="near-places__list places__list">
                {OffersListWithActiveItem}
              </div>
            </section>
          </div>
        </main>
      </div>

    </React.Fragment>;
  }
}

OfferDetails.propTypes = {
  activeOffer: PropTypes.object,
  isAuthorizationRequired: PropTypes.bool,
  user: PropTypes.object,
  cityName: PropTypes.string,
  changeActiveOffer: PropTypes.func,
  offers: PropTypes.array
};


