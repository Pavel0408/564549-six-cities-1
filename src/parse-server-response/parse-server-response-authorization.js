export const parseAuthorizationResponse = (response) => {
  const data = response.data;

  return {
    id: data[`id`],
    email: data[`email`],
    name: data[`name`],
    avatar: data[`avatar_url`],
    isPro: data[`is_pro`]
  };
};
