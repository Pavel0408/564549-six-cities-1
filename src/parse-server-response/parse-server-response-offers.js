const ParseServerResponseOffers = (data) => {
  return {
    name: data.name,
    image: data.image,
    price: data.price,
    rating: data.rating,
    isPremium: data.isPremium,
    isFavorite: data.isFavorite,
    coordinates: data.coordinates,
    city: data.city,
    allImages: data.allImages,
    type: data.type,
    bedrooms: data.bedrooms,
    maxAdults: data.maxAdults,
    goods: data.goods,
    host: {
      id: data.host.id,
      isPro: data.host.isPro,
      name: data.host.name,
      avatar: data.host.avatar
    },
    description: data.description,
    zoom: data.zoom,
    id: data.id
  };
};

const parseOfferFromServer = (data) => {
  return new ParseServerResponseOffers({
    name: data[`title`],
    image: data[`preview_image`],
    price: data[`price`],
    rating: Math.round(data[`rating`] * 2),
    isPremium: data[`is_premium`],
    isFavorite: data[`is_favorite`],
    coordinates: [data[`location`][`latitude`], data[`location`][`longitude`]],
    city: data[`city`][`name`],
    allImages: data[`images`],
    type: data [`type`],
    bedrooms: data[`bedrooms`],
    maxAdults: data[`max_adults`],
    goods: data[`goods`],
    host: {
      id: data[`host`][`id`],
      isPro: data[`host`][`is_pro`],
      name: data[`host`][`name`],
      avatar: data[`host`][`avatar_url`]
    },
    description: data[`description`],
    zoom: data[`location`][`zoom`],
    id: data[`id`]
  });
};

export const parseServerResponseOffers = (response) => {
  return response.data.map(parseOfferFromServer);
};

