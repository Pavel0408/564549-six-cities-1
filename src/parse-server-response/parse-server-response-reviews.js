const parseReview = (data) => {
  return {
    id: data[`id`],
    userId: data[`user`][`id`],
    userIsPro: data[`user`][`is_pro`],
    userName: data[`user`][`name`],
    userAvatar: data[`user`][`avatar_url`],
    rating: data[`rating`],
    message: data[`comment`],
    date: new Date(data[`date`])
  };
};

export const parseServerResponseReviews = (response) => {
  return response.data.map(parseReview);
};
