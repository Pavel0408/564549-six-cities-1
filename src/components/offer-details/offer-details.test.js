import React from 'react';
import renderer from 'react-test-renderer';


import {parseServerResponseOffers} from "../../parse-server-response/parse-server-response-offers";
import {BrowserRouter} from "react-router-dom";
import {OfferDetails} from "./offer-details";

const mockServerOffers = {
  data: [
    {
      'id': 1,
      'city': {
        'name': `Amsterdam`,
        'location': {
          'latitude': 52.370216,
          'longitude': 4.895168,
          'zoom': 10
        }
      },
      'preview_image': `img/1.png`,
      'images': [`img/1.png`, `img/2.png`],
      'title': `Beautiful & luxurious studio at great location`,
      'is_favorite': false,
      'is_premium': false,
      'rating': 4.8,
      'type': `apartment`,
      'bedrooms': 3,
      'max_adults': 4,
      'price': 120,
      'goods': [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
      'host': {
        'id': 3,
        'is_pro': true,
        'name': `Angelina`,
        'avatar_url': `img/1.png`
      },
      'description': `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
      'location': {
        'latitude': 52.35514938496378,
        'longitude': 4.673877537499948,
        'zoom': 8
      }
    }
  ]

};

const offerMock = parseServerResponseOffers(mockServerOffers);
const fetchReviews = () => {};

describe(`testing OfferDetails render`, () => {
  it(`OfferDetails is render correctly`, () => {
    const tree = renderer.create(
        <BrowserRouter>
          <OfferDetails
            activeOffer={offerMock[0]}
            isAuthorizationRequired={false}
            user={null}
            offers={offerMock}
            fetchReviews={fetchReviews}
            cityName={`Amsterdam`}
            reviews={[]}
          />
        </BrowserRouter>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
