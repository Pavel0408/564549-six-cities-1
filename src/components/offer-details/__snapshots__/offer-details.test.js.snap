// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing OfferDetails render OfferDetails is render correctly 1`] = `
<div>
  <div
    style={
      Object {
        "display": "none",
      }
    }
  >
    <svg
      xmlns="http://www.w3.org/2000/svg"
    >
      <symbol
        id="icon-arrow-select"
        viewBox="0 0 7 4"
      >
        <path
          clipRule="evenodd"
          d="M0 0l3.5 2.813L7 0v1.084L3.5 4 0 1.084V0z"
          fillRule="evenodd"
        />
      </symbol>
      <symbol
        id="icon-bookmark"
        viewBox="0 0 17 18"
      >
        <path
          d="M3.993 2.185l.017-.092V2c0-.554.449-1 .99-1h10c.522 0 .957.41.997.923l-2.736 14.59-4.814-2.407-.39-.195-.408.153L1.31 16.44 3.993 2.185z"
        />
      </symbol>
      <symbol
        id="icon-star"
        viewBox="0 0 13 12"
      >
        <path
          clipRule="evenodd"
          d="M6.5 9.644L10.517 12 9.451 7.56 13 4.573l-4.674-.386L6.5 0 4.673 4.187 0 4.573 3.549 7.56 2.483 12 6.5 9.644z"
          fillRule="evenodd"
        />
      </symbol>
    </svg>
  </div>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height={41}
              src="/img/logo.svg"
              width={81}
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/login"
                onClick={[Function]}
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  className="header__login"
                >
                  Sign in
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="img/1.png"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="img/2.png"
            />
          </div>
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              Beautiful & luxurious studio at great location
            </h1>
            <button
              className="property__bookmark-button button"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="property__bookmark-icon"
                height={33}
                style={Object {}}
                width={31}
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              5
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            >
              Entire place
            </li>
            <li
              className="property__feature property__feature--bedrooms"
            >
              3
               Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 
              4
               adults
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              120
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            >
              <li
                className="property__inside-item"
              >
                Heating
              </li>
              <li
                className="property__inside-item"
              >
                Kitchen
              </li>
              <li
                className="property__inside-item"
              >
                Cable TV
              </li>
              <li
                className="property__inside-item"
              >
                Washing machine
              </li>
              <li
                className="property__inside-item"
              >
                Coffee machine
              </li>
              <li
                className="property__inside-item"
              >
                Dishwasher
              </li>
            </ul>
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper"
              >
                <img
                  alt="Host avatar"
                  className="property__avatar user__avatar"
                  height={74}
                  src="/img/1.png"
                  width={74}
                />
              </div>
              <span
                className="property__user-name"
              >
                Angelina
              </span>
              <span
                className="property__user-status"
              >
                Pro
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              >
                A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
              </p>
            </div>
          </div>
          <section
            className="property__reviews reviews"
          >
            <h2
              className="reviews__title"
            >
              Reviews · 
              <span
                className="reviews__amount"
              >
                0
              </span>
            </h2>
            <ul
              className="reviews__list"
            />
          </section>
        </div>
      </div>
      <section
        className="property__map map"
        style={
          Object {
            "backgroundImage": "none",
          }
        }
      >
        <div
          id="map"
          style={
            Object {
              "backgroundImage": "none",
              "height": "100%",
            }
          }
        />
      </section>
    </section>
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
        <div
          className="near-places__list places__list"
        />
      </section>
    </div>
  </main>
</div>
`;
